### ISSUE TITLE
Difference between local shap values coming from global and from the "shap_values"-method

### ISSUE THREAD

**Issue by rainergo** [shap/shap#2101]
Hi, 
I have a tree explanation object containing ".data", ".values" and ".base_values". If I extract the shapley values from this object for a single data row ("local" value), I get the shapley values for the 32 features of my single data row.

Now, if I run the method "shap_values" on the explainer object and pass to it the feature data for that particular row, I also get the shapley values for the same row as above.

In my understanding, the shapley values extracted from the explanation object and the shapley values from the "shap_values"-method run on the explainer object should be the same. 

But they are not. Very often (not always), there is a difference for only ONE feature shapley value. In my case, the shapley values for 31 features are exactly the same (for the two approaches), but most often there is a difference in ONE feature shapley value. I tried this on different data sets ... .

Why is that? And is this a bug?

**Comment by c56pony** [comment]
I was not able to reproduce that error in my environment. Can you please tell me the code to reproduce it?
My code and output is as follows.
```
from sklearn.model_selection import train_test_split
import xgboost as xgb
import shap
import numpy as np

X, y = shap.datasets.adult()
train_x, valid_x, train_y, valid_y = train_test_split(X, y, test_size=0.25, random_state=7)
dtrain = xgb.DMatrix(train_x, label=train_y)
dvalid = xgb.DMatrix(valid_x, label=valid_y)

params = {
    "objective": "binary:logistic",
    "eval_metric": "auc",
    "eta": 0.7
}
model = xgb.train(params, dtrain, evals = [(dvalid, "valid")])

explainer = shap.TreeExplainer(model)
shap_values = explainer(X)
shap_values2 = explainer.shap_values(X)
np.allclose(shap_values.values, shap_values2)
```
outputs `True`

**Comment by LasseVDHeydtQC** [comment]
Having the same issue, the values are just different... Could it be that the two methods handle categorical dtype differently?

**Comment by firobeid** [comment]
You guys should index based on the second axis to get the Shap's for each feature alone:
`
shap_values.values[:,feature_index]
`
where feature_index is the index in the train set of the feature in the column.
